File: midonet-api/src/main/java/org/midonet/api/network/rest_api/VtepResource.java
/*
 * Copyright 2014 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.api.network.rest_api;

import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.List;

import javax.annotation.security.RolesAllowed;
import javax.validation.Validator;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.SecurityContext;
import javax.ws.rs.core.UriInfo;

import org.midonet.api.ResourceUriBuilder;
import org.midonet.cluster.rest_api.VendorMediaType;
import org.midonet.api.auth.AuthRole;
import org.midonet.api.network.VTEPPort;
import org.midonet.cluster.rest_api.BadRequestHttpException;
import org.midonet.cluster.rest_api.ConflictHttpException;
import org.midonet.cluster.rest_api.GatewayTimeoutHttpException;
import org.midonet.cluster.rest_api.NotFoundHttpException;
import org.midonet.api.rest_api.ResourceFactory;
import org.midonet.api.rest_api.RestApiConfig;
import org.midonet.api.vtep.VtepClusterClient;
import org.midonet.cluster.DataClient;
import org.midonet.cluster.data.host.Host;
import org.midonet.cluster.data.vtep.VtepNotConnectedException;
import org.midonet.cluster.data.vtep.model.PhysicalSwitch;
import org.midonet.cluster.rest_api.conversion.VTEPDataConverter;
import org.midonet.cluster.rest_api.models.VTEP;
import org.midonet.cluster.rest_api.validation.MessageProperty;
import org.midonet.midolman.serialization.SerializationException;
import org.midonet.midolman.state.NoStatePathException;
import org.midonet.midolman.state.NodeNotEmptyStateException;
import org.midonet.midolman.state.StateAccessException;
import org.midonet.midolman.state.StatePathExistsException;
import org.midonet.packets.IPv4Addr;

import com.google.inject.Inject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static scala.collection.JavaConversions.asJavaCollection;

import static org.midonet.cluster.rest_api.validation.MessageProperty.VTEP_EXISTS;
import static org.midonet.cluster.rest_api.validation.MessageProperty.VTEP_HAS_BINDINGS;
import static org.midonet.cluster.rest_api.validation.MessageProperty.VTEP_HOST_IP_CONFLICT;
import static org.midonet.cluster.rest_api.validation.MessageProperty.VTEP_NOT_FOUND;
import static org.midonet.cluster.rest_api.validation.MessageProperty.getMessage;

public class VtepResource extends AbstractVtepResource {
    private final static Logger log =
        LoggerFactory.getLogger(VtepResource.class);

    @Inject
    public VtepResource(RestApiConfig config, UriInfo uriInfo,
                        SecurityContext context, Validator validator,
                        DataClient dataClient, ResourceFactory factory,
                        VtepClusterClient vtepClient) {
        super(config, uriInfo, context, validator,
              dataClient, factory, vtepClient);
    }

    @POST
    @RolesAllowed({AuthRole.ADMIN})
    @Consumes({VendorMediaType.APPLICATION_VTEP_JSON,
               MediaType.APPLICATION_JSON})
    public Response create(VTEP vtep)
            throws SerializationException, StateAccessException {

        validate(vtep);

        if (!dataClient.tunnelZonesExists(vtep.tunnelZoneId)) {
            throw new BadRequestHttpException(
                getMessage(MessageProperty.TUNNEL_ZONE_ID_IS_INVALID));
        }

        org.midonet.cluster.data.VTEP dataVtep = VTEPDataConverter.toData(vtep);
        List<InetAddress> vtepIps = new ArrayList<>();

        // Verify there is no conflict between hosts and the VTEP IPs.
        try {
            PhysicalSwitch ps = vtepClient.getPhysicalSwitch(
                IPv4Addr.apply(vtep.managementIp),
                vtep.managementPort);

            // Check all management and tunnel IPs configured for the physical
            // switch.
            addIpsToList(vtepIps, ps.mgmtIpStrings());
            addIpsToList(vtepIps, ps.tunnelIpStrings());

        } catch(GatewayTimeoutHttpException | VtepNotConnectedException e) {
            log.warn("Cannot verify conflicts between hosts and VTEP IPs "
                     + " because VTEP {}:{} is not accessible",
                     vtep.managementIp, vtep.managementPort);
        }

        addIpToList(vtepIps, vtep.managementIp);

        for (Host host : dataClient.hostsGetAll()) {
            for (InetAddress ip : host.getAddresses()) {
                if (vtepIps.contains(ip)) {
                    throw new ConflictHttpException(
                        getMessage(VTEP_HOST_IP_CONFLICT, ip));
                }
            }
        }

        try {
            dataClient.vtepCreate(dataVtep);
            return Response.created(ResourceUriBuilder.getVtep(
                    getBaseUri(), dataVtep.getId().toString())).build();
        } catch(StatePathExistsException ex) {
            throw new ConflictHttpException(
                ex, getMessage(VTEP_EXISTS, vtep.managementIp));
        }
    }

    @GET
    @RolesAllowed({AuthRole.ADMIN})
    @Path("{ipAddr}")
    @Produces({VendorMediaType.APPLICATION_VTEP_JSON,
               MediaType.APPLICATION_JSON})
    public VTEP get(@PathParam("ipAddr") String ipAddrStr)
            throws StateAccessException, SerializationException {

        IPv4Addr ipAddr = parseIPv4Addr(ipAddrStr);
        org.midonet.cluster.data.VTEP dataVtep =
                vtepClient.getVtepOrThrow(ipAddr, false);
        return toApiVtep(dataVtep);
    }

    @GET
    @RolesAllowed({AuthRole.ADMIN})
    @Produces({VendorMediaType.APPLICATION_VTEP_COLLECTION_JSON,
               MediaType.APPLICATION_JSON})
    public List<VTEP> list()
            throws StateAccessException, SerializationException {
        List<org.midonet.cluster.data.VTEP> dataVteps = dataClient.vtepsGetAll();
        List<VTEP> vteps = new ArrayList<>(dataVteps.size());
        for (org.midonet.cluster.data.VTEP dataVtep : dataVteps) {
            vteps.add(toApiVtep(dataVtep));
        }
        return vteps;
    }

    @DELETE
    @RolesAllowed({AuthRole.ADMIN})
    @Path("{ipAddr}")
    public void delete(@PathParam("ipAddr") String ipAddrStr)
            throws SerializationException, StateAccessException {
        try {
            vtepClient.deleteVtep(parseIPv4Addr(ipAddrStr));
        } catch (NoStatePathException ex) {
            throw new NotFoundHttpException(getMessage(
                    VTEP_NOT_FOUND, ipAddrStr));
        } catch (NodeNotEmptyStateException ex) {
            throw new BadRequestHttpException(getMessage(
                    VTEP_HAS_BINDINGS, ipAddrStr));
        }
    }

    @GET
    @RolesAllowed({AuthRole.ADMIN})
    @Path("{ipAddr}" + ResourceUriBuilder.PORTS)
    @Produces({VendorMediaType.APPLICATION_VTEP_PORT_COLLECTION_JSON,
               MediaType.APPLICATION_JSON})
    public List<VTEPPort> listPorts(@PathParam("ipAddr") String ipAddrStr)
            throws SerializationException, StateAccessException {
        return vtepClient.listPorts(parseIPv4Addr(ipAddrStr));
    }

    @Path("/{ipAddr}" + ResourceUriBuilder.BINDINGS)
    public VtepBindingResource getVtepBindingResource(
        @PathParam("ipAddr") String ipAddrStr) {
        return factory.getVtepBindingResource(ipAddrStr);
    }

    private VTEP toApiVtep(org.midonet.cluster.data.VTEP dataVtep) {
        PhysicalSwitch ps;
        try {
            ps = vtepClient.getPhysicalSwitch(dataVtep.getId(),
                                              dataVtep.getMgmtPort());
        } catch (Exception ex) {
            ps = null;
        }

        VTEP apiVtep = VTEPDataConverter.fromData(dataVtep, ps);
        apiVtep.setBaseUri(getBaseUri());
        return apiVtep;
    }

    private static void addIpToList(List<InetAddress> list, String ip) {
        try {
            InetAddress address = InetAddress.getByName(ip);
            list.add(address);
        } catch (UnknownHostException e) {
        }
    }

    private static void addIpsToList(List<InetAddress> list,
                                  scala.collection.Iterable<String> ips) {
        for (String ip : asJavaCollection(ips)) {
            addIpToList(list, ip);
        }
    }
}

File: midonet-cluster/src/main/scala/org/midonet/cluster/services/rest_api/resources/VtepResource.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.services.rest_api.resources

import java.util.UUID
import javax.ws.rs._
import javax.ws.rs.core.MediaType.APPLICATION_JSON

import scala.collection.JavaConversions._
import scala.concurrent.Future

import com.google.inject.Inject
import com.google.inject.servlet.RequestScoped

import org.midonet.cluster.data.storage.NotFoundException
import org.midonet.cluster.models.State.VtepConnectionState._
import org.midonet.cluster.models.State.{VtepConfiguration, VtepConnectionState}
import org.midonet.cluster.rest_api._
import org.midonet.cluster.rest_api.annotation._
import org.midonet.cluster.rest_api.models.TunnelZone.TunnelZoneType
import org.midonet.cluster.rest_api.models.Vtep.ConnectionState._
import org.midonet.cluster.rest_api.models.{TunnelZone, Vtep}
import org.midonet.cluster.rest_api.validation.MessageProperty._
import org.midonet.cluster.services.rest_api.MidonetMediaTypes._
import org.midonet.cluster.services.rest_api.resources.MidonetResource.{NoOps, Ops, ResourceContext}
import org.midonet.cluster.services.vxgw.data.VtepStateStorage._
import org.midonet.packets.{IPAddr, IPv4Addr}
import org.midonet.southbound.vtep.{OvsdbVtepConnectionProvider, OvsdbVtepDataClient}
import org.midonet.util.reactivex._

@ApiResource(version = 1)
@Path("vteps")
@RequestScoped
@AllowGet(Array(APPLICATION_VTEP_JSON_V2,
                APPLICATION_JSON))
@AllowList(Array(APPLICATION_VTEP_COLLECTION_JSON_V2,
                 APPLICATION_JSON))
@AllowCreate(Array(APPLICATION_VTEP_JSON_V2,
                   APPLICATION_JSON))
@AllowDelete
class VtepResource @Inject()(resContext: ResourceContext,
                             cnxnProvider: OvsdbVtepConnectionProvider)
    extends MidonetResource[Vtep](resContext) {

    @Path("{id}/bindings")
    def bindings(@PathParam("id") vtepId: UUID): VtepBindingResource = {
        new VtepBindingResource(vtepId, resContext, cnxnProvider)
    }

    @Path("{id}/ports")
    def ports(@PathParam("id") vtepId: UUID): VtepPortResource = {
        new VtepPortResource(vtepId, resContext, cnxnProvider)
    }

    protected override def getFilter(vtep: Vtep): Future[Vtep] = {
        initVtep(vtep)
    }

    protected override def listFilter(vteps: Seq[Vtep]): Future[Seq[Vtep]] = {
        Future.sequence(for (vtep <- vteps) yield initVtep(vtep))
    }

    protected override def createFilter(vtep: Vtep): Ops = {

        throwIfViolationsOn(vtep)

        // Validate the tunnel zone.
        if (vtep.tunnelZoneId eq null) {
            val msg = getMessage(TUNNEL_ZONE_ID_IS_INVALID)
            throw new BadRequestHttpException(msg)
        }

        getResource(classOf[TunnelZone], vtep.tunnelZoneId) recover {
            // Validate the tunnel zone exists.
            case e: NotFoundException =>
                val msg = getMessage(TUNNEL_ZONE_ID_IS_INVALID)
                throw new BadRequestHttpException(msg)
        } flatMap { tunnelZone =>
            // Validate this is a VTEP tunnel-zone
            if (tunnelZone.`type` != TunnelZoneType.vtep) {
                val msg = getMessage(TUNNEL_ZONE_NOT_VTEP)
                throw new BadRequestHttpException(msg)
            }

            listResources(classOf[Vtep])
        } map { vteps =>
            // Validate there is no conflict with existing VTEPs.
            for (v <- vteps if v.managementIp == vtep.managementIp) {
                val msg = getMessage(VTEP_EXISTS, vtep.managementIp)
                throw new ConflictHttpException(msg)
            }
        } getOrThrow

        // Verify there is no conflict between hosts and the VTEP IPs.
        val mgmtIp = IPv4Addr.fromString(vtep.managementIp)
        val cnxn = cnxnProvider.get(mgmtIp, vtep.managementPort)
        val client = OvsdbVtepDataClient(cnxn)

        try {
            val vtepIps = client.connect() flatMap { _ =>
                client.physicalSwitch
            } map {
                case Some(physicalSwitch) =>
                    (physicalSwitch.mgmtIpStrings.map(IPv4Addr.fromString) ++
                    physicalSwitch.tunnelIpStrings.map(IPv4Addr.fromString) ++
                    Seq(IPv4Addr(vtep.managementIp))).toSet
                case None =>
                    throw new ServiceUnavailableHttpException(
                        s"Cannot connect to VTEP at ${vtep.managementIp}:" +
                        s"${vtep.managementPort} because the physical switch " +
                        "is not configured")
            } getOrThrow

            // Typically we'd go to backend.store directly, however we care
            // about host ips and addresses so we delegate on HostResource to
            // do the composition for us.
            val hostResource = new HostResource(resContext)
            val hostIps =
                hostResource.list(APPLICATION_HOST_COLLECTION_JSON_V3)
                            .flatMap { _.addresses.map(IPAddr.fromString) }
                            .toSet

            val commonIps = vtepIps.find(hostIps.contains)
            if (commonIps.nonEmpty) {
                val msg = getMessage(VTEP_HOST_IP_CONFLICT, commonIps.head)
                throw new ConflictHttpException(msg)
            }
        } finally {
            client.close()
        }

        vtep.create()
        NoOps
    }

    protected override def deleteFilter(id: String): Ops = {
        getResource(classOf[Vtep], id) map { vtep =>
            // Validate the VTEP has no bindings.
            if (vtep.bindings.size() > 0) {
                val msg = getMessage(VTEP_HAS_BINDINGS, vtep.managementIp)
                throw new BadRequestHttpException(msg)
            }
        } getOrThrow

        NoOps
    }

    private def initVtep(vtep: Vtep): Future[Vtep] = {
        getConfiguration(vtep.id) flatMap { config =>
            vtep.name = config.getName
            vtep.description = config.getDescription
            vtep.tunnelIpAddrs = config.getTunnelAddressesList.map(_.getAddress)
            getConnectionState(vtep.id)
        } map { connectionState =>
            vtep.connectionState = connectionState match {
                case VTEP_DISCONNECTED => disconnected
                case VTEP_CONNECTED => connected
                case VTEP_ERROR => error
            }
            vtep
        }
    }

    private def getConfiguration(vtepId: UUID): Future[VtepConfiguration] = {
        resContext.backend.stateStore.getVtepConfig(vtepId).asFuture
    }

    private def getConnectionState(vtepId: UUID): Future[VtepConnectionState] = {
        resContext.backend.stateStore.getVtepConnectionState(vtepId).asFuture
    }

}
